crypto/encoding/codec.go

```diff
@@ -32,6 +32,12 @@ func PubKeyToProto(k crypto.PubKey) (pc.PublicKey, error) {
 				Secp256K1: k,
 			},
 		}
+	case secp256k1.PubKeyOld:
```

Comment:
Good catch.
When I set up this patch, I ran all the nightly `e2e` tests and they were passing.
I just looked at the usages of `PubKeyToProto` and `PubKeyFromProto`, and there are two usages of this that may justify also adapting `PubKeyFromProto`:
* external signer: is this used by Heimdall node operators, if yes, @lasarojc's suggestion must be implemented
* in some ABCI calls the App (SDK) has to fill up some fields with (an equivalent of) `PubKeyFromProto`. If we're not sure what Heimdall's SDK is using, it's probably a good idea to implement @lasarojc's suggestion

Reply:
Wouldn't `PublicKey_Secp256K1` encompass both old and new pub key types ? I guess it might be clearer to define another proto definition for the old type but at the end it'd be just a sequence of bytes. 
Besides, from the perspective of this fork, I believe we shouldn't need to make any change (since we're only supporting eth pub keys) but in upstream comet I imagine it would have to handle both pub key sizes.

Reply:
If an operator didn't update it's JSON files to go from "tendermint/PubKeySecp256k1" to "comet/PubKeySecp256k1Uncompressed" (this will be possible

Remember there will be a transition period: the release where peppermint keeps this code and operators are encouraged to migrate from "tendermint/PubKeySecp256k1" to "comet/PubKeySecp256k1Uncompressed" in their JSON files.

During this period, it is possible (and it should be supported) to have operators that still have occurrences of "tendermint/PubKeySecp256k1" in their JSON files. If the peppermint code reads those JSON files (for instance an external signer, see my comment above) then, when function `PubKeyFromProto` is called, the type passed will be `secp256k1.PubKeyOld`. So I think @lasarojc's initial comment make sense, and we should add `secp256k1.PubKeyOld` to the switch in function `PubKeyFromProto`.

Reply:
Hmm, the cases in `PubKeyFromProto` are based on the curve type of the public key. Since `PublicKey_Secp256K1` is just a sequence of bytes generated from the proto definition

```
message PublicKey {
  option (gogoproto.compare) = true;
  option (gogoproto.equal)   = true;

  oneof sum {
    bytes ed25519   = 1;
    bytes secp256k1 = 2;
  }
}
```
, I'm not sure whether there is a way to segregate between the old and new format of key. Maybe I'm missing something here. 
Let's have a brief chat about this in today's meeting.

Reply:
Thanks @Raneet10 for your comment, and for the discussion yesterday. I've spent some time further investigating this and this is what I found out:

* Firstly, the root problem we are having here is that, when I helped porting secp256k1 from old peppermint, I only dealt with the "domain types" associated to `PubKey` defined in `crypto/secp256k1/secp256k1.go`. I didn't think of extending the corresponding protobuf tyes. For context, we call "domain types" the manually defined "mirror" types that correspond 1-to-1 to protobuf types.
* That's why you can't add an "OldSecp256k1" case here :-(
* Now, in the upstreamed changes (in order to support both compressed and uncompressed) the CometBFT team will need to add a new case in the `oneof sum` of `message PublicKey` for the uncompressed case (`PublicKey_Secp256K1Uncompressed`), since we will support both. In that version, a protobuf with `PublicKey_Secp256K1` will represent _only_ a compressed public key.
* Then, I re-visited all the code that calls `PubKeyFromProto`: it is mainly used as part of the protobuf validator struct.
* Unfortunately, CometBFT's state store is using the `ValidatorSet` protobuf type to store info about validators, which uses `message PublicKey`, e.g., [here](https://github.com/0xPolygon/cometbft/blob/v0.38.5-upstream/state/store.go#L559).
*  For now, this is not a problem. BUT, we will be in trouble when upstreaming these changes from CometBFT, as operators will have marshalled to `PublicKey_Secp256K1` and stored that in the state store (I mean, a CometBFT DB, not a JSON file!), making it more complex for operators to upgrade to a version that gets rid of this fork code: when unmarshalling info about past heights (e.g. validator sets), the state store would get a `PublicKey_Secp256K1` (so, compressed) and pass it on to the calling function.
* I'm sorry I overlooked this problem when helping forward-port peppermint's secp256k1 code. And I'm glad you insisted on this comment. We must add `PublicKey_Secp256K1Uncompressed` now, even if it doesn't seem needed ATM.
* Finally, bear in mind that we also need to adapt two places in (Heimdall's fork of) the SDK: [here](https://github.com/0xPolygon/cosmos-sdk/blob/v0.50.2/crypto/codec/cmt.go#L43) and [here](https://github.com/0xPolygon/cosmos-sdk/blob/v0.50.2/crypto/codec/cmt.go#L23)). Please note it down somewhere.

In summary (1) thanks for insisting here, (2) I will add the missing protobuf definition then I'm back from vacation (I'm off next week), (3) this is non-blocking for you folks, as nothing is breaking ATM (it _would_ be breaking when removing the secp256k1 changes from this fork.

Reply:
Makes sense, thanks for the detailed explanation! 
So, if I understood correctly, we _first_ need to add the definition for `PublicKey_Secp256K1Uncompressed` in this fork prior to upstreaming right ? I can take it up in the meantime you're away in case.

Reply:
Exactly.
With one extra "trick": in function `PubKeyToProto`, in both `case secp256k1.PubKeyOld` and `case secp256k1.PubKeyOld`, we use the new `PublicKey_Secp256K1Uncompressed`. IOW, we _never_ convert to `PublicKey_Secp256K1`.
Actually, I would even go further, and remove the `case *pc.PublicKey_Secp256K1` in `PubKeyToProto` (only `case *pc.PublicKey_Secp256K1Uncompressed` would stay), to make sure we NOT accepting the uncompressed from anywhere.

Reply:
Yes I was about to suggest the same! Although I guess it won't stay close to upstream but that shouldn't matter given we've already changed secp256k1 logic heavily. I'll try to take a stab at it in the meantime you're off. Thanks again for painstakingly looking into this, glad we came upon this!

Reply:
Made the respective changes on SDK: 0xPolygon/cosmos-sdk#6

Reply:
As mentioned during last week's meeting, I just finished doing the precautionary code traversal with this change on it and everything looks good to me.

Resolved